### PR TITLE
[GSoC] Sampling from external libraries for all RVs

### LINKED ISSUE #6160: Stats sampling using numpy/Theano
```
The stats module has a sampling mechanism to provide approximate solutions to statistical queries. This mechanism is relatively slow and 1000+ sample simulations of complex expressions can take several seconds. 

>>> X = Normal(0,1)
>>> Y = Exponential(10)
>>> Var(sin(X+Y)) # this hangs because the integral is tough
<Ctrl-C>
>>> Var(sin(X+Y), numsamples=1000) # this takes a few seconds 
0.420095049234496

NumPy also provides a sampling mechanism in its random module. If we could sample large arrays out of the underlying distributions and then push those large arrays through the sympy expressions we could compute approximate solutions much more quickly. 

>>> import numpy
>>> xs = numpy.random.normal(0,1, 1000)
>>> ys = numpy.random.exponential(10, 1000)
>>> numpy.sin(xs+ys).var() # This takes 150 micro seconds
0.48987452637810897

So, can we use sympy to build up a complex statistical question and then use numpy to compute the result?
```

Original issue for #6160: http://code.google.com/p/sympy/issues/detail?id=3061
Original author: https://code.google.com/u/109882876523836932473/
Original owner: https://code.google.com/u/109882876523836932473/

### LINKED ISSUE #7358: Use scipy.stats when sampling from sympy.stats distributions
The `sample` operation samples a random expression. 

```
>>> X = Normal('X', 0, 1)
>>> sample(100 + X)
100.325245
```

This depends on `*Distribution.sample` methods for each type of distribution (like `NormalDistribution`.)  These methods often call out to a sampling method in the `random` stdlib module.  When these methods do not exist we attempt to create one analytically by computing the CDF of the distribution.

In some cases the distribution is not found in `random` and the computation of the CDF is difficult.  When these two happen together it breaks our ability to sample.

One way to resolve this issue is to depend on the `scipy.stats` module instead of `random`, which is generally far more comprehensive.  This dependence would need to be shielded in case `scipy` wasn't present on the user's machine.

If someone wanted to fix this they should look at `sympy/stats/crv_types.py`, take a look at `GammaDistribution.sample` (works) and then try to implement the same thing for `GammaInverseDistribution.sample` (does not yet exist) using `scipy.stats.invgamma`.  Once that works there are several more!

### LINKED ISSUE #8478: Sampling from poisson distribution fails
The following:

``` python
>>> p = Poisson('p', Rational(1, 10))
>>> sample(p)
```

raises

``` pytb
Traceback (most recent call last):
  File "<ipython-input-5-16980c83ac2c>", line 1, in <module>
    sample(p)
  File "sympy/stats/rv.py", line 783, in sample
    return next(sample_iter(expr, condition, numsamples=1))
  File "sympy/stats/rv.py", line 879, in sample_iter_subs
    d = ps.sample()  # a dictionary that maps RVs to values
  File "sympy/stats/drv.py", line 106, in sample
    return {self.value: self.distribution.sample()}
  File "sympy/stats/drv.py", line 31, in sample
    icdf = self._inverse_cdf_expression()
  File "sympy/core/cache.py", line 91, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/core/compatibility.py", line 872, in wrapper
    result = user_function(*args, **kwds)
  File "sympy/stats/drv.py", line 47, in _inverse_cdf_expression
    raise NotImplementedError("Could not invert CDF")
NotImplementedError: Could not invert CDF
```

Inverting the CDF is only one way of sampling from a discrete distribution, and I am not even sure it is the most efficient.

### LINKED ISSUE #16942: [Discussion] Exporting expression of RVs to External Libraries
There are 2 objective to this topic:
1. Use external libraries to sample from simple distributions. Here we can use `numpy`, `scipy`, `pymc3` and `stan` for doing sampling for simple distributions. Eg.
```python
>>> # NumPy
>>> X = Normal('X', 0, 1)
>>> c = export_normal(X, numpy=True, size=3)
>>> c['numpy']
array([-0.85648314, -1.67294285,  1.32458402])

>>> # PyMC3
>>> X = Normal('X', 0, 1)
>>> c = export_normal(X, pymc3=True, size=3)
>>> c['pymc3'][:]['X']
array([ 0.70263423, -0.59543018, -0.36750535])
```
2. Use external libraries for doing inference. When performing probabilistic inference, often we don't have the final distribution in closed form. In such situations, we can use sampling-based methods given in `PyMC3` and `stan` to get the sample (refer [pymc3/sampling](https://docs.pymc.io/api/inference.html) and [stan/sampling](https://mc-stan.org/rstan/reference/stanmodel-method-sampling.html)). Here I propose that we should be able to sample for compound distributions for which we can't find the final distribution in closed form. 
---
**Note from my proposal:** I couldn't find a general method to compute the density of compound RV (I believe that such general formula doesn't exist, because it would require integrating out at least one parameter, e.g. non-conjugate prior in machine learning). But for some instances, it is possible as given [here](https://en.wikipedia.org/wiki/Compound_probability_distribution#Examples).

---
Therefore, whenever we are not able to simplify the compound distribution, we should give an option to the user to export the RV to `pymc3` or `stan` for inference. *This will require us to break down the distribution into priors and likelihood distribution.* Eg.
```python3
# Suppose we have a normal distribution RV `X` with hyperparameter `mu` and `sigma`
# distributed uniformly in [10, 20]. Since we cannot find the exact distribution of `X`,
# we can export `X` to pymc3 to do inference by sampling.
# For this we have to find the distributions of `mu` and `sigma` and
# then run `pymc3` model to do inference.
>>> mu = Uniform('mu', 10, 20)
>>> sigma = Uniform('sigma', 10, 20)
>>> X = Normal('X', mu, sigma)
>>> E(X)
NotImplementedError: 
Expectations of expression with unindexed joint random symbols cannot
be calculated yet.

>>> trace = export_compound_normal(X, pymc3=True, size=100)
>>> import numpy as np
>>> np.mean(trace[:]['X'])
15.01604506092257
```

**Implementation**
First case:
```python3
def export_normal(rv, *args, **kwargs):
    if len(kwargs) == 0:
        raise ValueError('Please enter the library to'
                         ' which the random is to be exported.')
    objs = {}
    name = rv.args[0].name
    mu, sigma = rv.pspace.distribution.args
    if isinstance(mu, Symbol) or isinstance(sigma, Symbol):
        raise ValueError('Symbolic parameters cannot be '
                        ' exported.')
    mu, sigma = int(mu), int(sigma)
    if kwargs.get('numpy', False):
        objs['numpy'] = np.random.normal(mu, sigma, size=kwargs.get('size', 1))
    if kwargs.get('pymc3', False):
        pymc3_model = pm.Model()
        with pm.Model() as model:
            X = pm.Normal(name, mu, sigma)
            objs['pymc3'] = pm.sample(size, chains=1)
    return objs
```
Second case: For this I will send a PR with basic prototype for discussion.

I have gone through `lambdify.py` in `sympy/utilities/lambdify.py` and will follow a similar path for `Probability export`.

Ping @Upabjojr @sidhantnagpal @czgdp1807

### LINKED ISSUE #19061: Inconsistent API of sampling from random variables
While working on sampling from external libraries in PR #18754, I see that `size` and `numsamples` arguments in different functions mean the same and has created an inconsistency of using which of them.

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1020-L1040
In `sample` we have `size` as public API, while returning `numsamples` is by default kept as 1.

The following samplings have `size` and `numsamples` both as the argument.
https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1043

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1130

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1177

The following samplings have only `numsamples` as an argument.
https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1207

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1243

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1265

These lead to confusion and fails many times as:
```
>>> from sympy.stats import *
>>> X=Normal('X',1,2)
>>> sample(X)     # works fine
-1.57391020015325

>>> sample(X, size=3)   # works fine
[1.69231493540159, 1.69231493540159, 1.69231493540159]

>>> Y=Normal('Y',3,4)
>>> sample(X+Y, size=3)  # fails : expected 3
8.01476115427784

>>> sample(X, numsamples=3)  # fails : expected 3
-0.132248747800019

>>> sample(X+Y, numsamples=3)  # fails : expected 3
3.62703730515697

>>> list(sample_iter(X, size=3, numsamples=2))  # unexpected output(neither 2 nor 3)
[[0.310554545718285, 0.310554545718285, 0.310554545718285], [1.27257976990205, 1.27257976990205, 1.27257976990205]]

>>> list(sample_iter(X, numsamples=3))
[-1.46157892264006, 4.01242293224973, -2.73175071462870]

>>> sampling_density(X, size=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/smit/Smit Lunagariya/sympy/sympy/stats/rv.py", line 1279, in sampling_density
    results[result] = results.get(result, 0) + 1
TypeError: unhashable type: 'list'

>>> sampling_density(X, numsamples=3) # we should expect a list of 3
{-2.61191742946724: 1, 4.20263875820254: 1, 3.97981174226570: 1}
```
Therefore, to prevent this confusion and failure, we can have only one argument throughout to maintain uniformity and consistency across all sampling methods.
Related issue #16942

### PR DESCRIPTION
**Pull request by Smit-create**
Sampling from External Libraries of sympy random variables
#### References to other Issues or PRs

FIxes #19061
Fixes #16942
FIxes #8478
Fixes #6160
FIxes #7358
#### Brief description of what is fixed or changed
Initially, as decided in #18754, there is some improvement in the design as previously all the external sampling was handled by a single class. Instead, there will be separate classes for each library to avoid load on a single class. Also, this PR completes the whole sampling of stats from external libraries including CRV, FRV, and DRV.

#### Release Notes

* stats
   * Added Sampling from external libraries for all the random variables of sympy

ping @czgdp1807 @Upabjojr @jmig5776

### CODE REVIEW

File: sympy/stats/tests/test_discrete_rv.py [hunk 1]
```diff
@@ -284,49 +284,34 @@ def test_product_spaces():
...
+def test_sample_scipy():
+    p = S(2)/3
+    x = Symbol('x', integer=True, positive=True)
+    pdf = p*(1 - p)**(x - 1) # pdf of Geometric Distribution
     distribs_scipy = [
+        DiscreteRV(x, pdf, set=S.Naturals),
         Geometric('G', 0.5),
         Logarithmic('L', 0.5),
+        # NegativeBinomial('N', 5, 0.2), # uncomment after lambdify handles binomial
         Poisson('P', 1),
```
Comment (Smit-create) [inline sympy/stats/tests/test_discrete_rv.py:296]: I get an error with `NegativeBinomial` as :
```
File "/home/smit/Smitlunagariya/sympy/sympy/stats/drv.py", line 38, in _pmf
    return handmade_pmf(x)
  File "<lambdifygenerated-20>", line 2, in _lambdifygenerated
    return (0.32768*0.2**Dummy_50*binomial(Dummy_50 + 4, Dummy_50))
NameError: name 'binomial' is not defined
```

**Comment by Smit-create** [comment]
This implementation works good for every CRV and FRV, I found only two failing examples in DRV:
1. NegativeBinomial-> `lambdify` error
2. Skellam-> Internal error in scipy..might be because of complex `pdf`

File: sympy/stats/crv.py [hunk 2]
```diff
@@ -25,8 +25,6 @@
         ProductDomain, PSpace, SinglePSpace, random_symbols, NamedArgsMixin)
 
 scipy = import_module('scipy')
-numpy = import_module('numpy')
-pymc3 = import_module('pymc3')
```
Comment (czgdp1807) [inline sympy/stats/crv.py:29]: Removal of these is fine as `pymc3` depends on `scipy`.

**Comment by czgdp1807** [comment]
I think @Upabjojr 's idea of using adaptors can work. Here's what I think,
Interface should accept random expression of sympy and sample it according to the base library selected and return only numpy arrays.
For class a base class, `Sampler` can be made. Samplers specific to libraries can inherit from base class. For example, `SciPySampler`. `Sampler` can transfer the calls to the methods of it's child class.
Each sampler can have implementations according to the libraries they are using for sampling purposes. This can be easy to extend, maintain and decoupled.

**Comment by czgdp1807** [comment]
Removing currently supported libraries have a downside that, users may have to modify their code a lot. And may be, one library can be having different algorithms(which can be better). So, removal of libraries narrows down use cases for the end user.

**Comment by Smit-create** [comment]
Then, the previous implementation looks fine if we think of providing multiple libraries.
We need not change that as the return type of all three of them is `numpy.ndarray`. So, the previous implementation looks good:
```python
>>> from sympy.stats import *
>>> N = Normal('N',0,1)
>>> resp=next(sample(N, library='pymc3', size=2))
>>> resp, resp.__class__
(array([0.28739264, 1.0677173 ]), <class 'numpy.ndarray'>)
>>> ress=next(sample(N, library='scipy', size=2))
>>> ress, ress.__class__
(array([-0.10553114, -1.55432616]), <class 'numpy.ndarray'>)
>>> resn=next(sample(N, library='numpy', size=2))
>>> resn, resn.__class__
(array([ 0.94539252, -0.68432651]), <class 'numpy.ndarray'>)
```
Then, the only change I suggest, is to use scipy custom distributions when `library` is `scipy` to avoid mechanically adding `scipy` rvs. Is it good?

**Comment by czgdp1807** [comment]
Well, the problem with current design is that everything is stored in a single sampling class and that makes it a bit hard to maintain.

**Comment by Smit-create** [comment]
So need to break into three classes?

**Comment by czgdp1807** [comment]
Yes, like a base sampler class which decides which one of it's child class should get the task of sampling.

**Comment by Smit-create** [comment]
> In addition, can we return `numpy.ndarray` instead of `iterator`?

Yes, Should I?

**Comment by Smit-create** [comment]
But then this may break the current API and the meaning of `sample_iter`

**Comment by czgdp1807** [comment]
Let's decide on the return type later on. First improve the design.

**Comment by Smit-create** [comment]
I have made a change for CRV, if this looks fine then will make similar to DRV and FRV

File: sympy/stats/crv.py [hunk 3]
```diff
@@ -150,42 +150,38 @@ def __call__(self, *args):
 
 
 class SampleExternalContinuous:
-    """Class consisting of the methods that are used to sample values of random
-    variables from external libraries."""
+    """Base Class to sample values of Continuous random variables from
+    external libraries."""
 
+    def __new__(cls, dist, size, library):
+        _sample_class = _get_sample_class[library]
```
Comment (czgdp1807) [inline sympy/stats/crv.py:157]:
```suggestion
        _sample_class = 'SampleContinuous' + upper(library[0]) + library[1:]
```

File: sympy/stats/crv.py [hunk 4]
```diff
@@ -150,42 +150,38 @@ def __call__(self, *args):
...
+        return _sample_class(dist, size)
```
Comment (czgdp1807) [inline sympy/stats/crv.py:158]: What if library isn't supported?

File: sympy/stats/crv.py [hunk 5]
```diff
@@ -274,6 +255,12 @@ def _sample_pymc3(cls, dist, size):
             cls.pymc3_rv_map[dist.__class__.__name__](dist)
             return pymc3.sample(size, chains=1, progressbar=False)[:]['X']
 
+_get_sample_class = {
+    'scipy': SampleContinuousScipy,
+    'pymc3': SampleContinuousPymc,
+    'numpy': SampleContinuousNumpy
+}
+
```
Comment (czgdp1807) [inline sympy/stats/crv.py:263]:
```suggestion
```

**Comment by czgdp1807** [comment]
I think you can apply the same design for discrete and finite variables.

File: sympy/stats/crv.py (hunk 4 above)
Comment (Smit-create) [reply sympy/stats/crv.py:158]: It will raise an error in distribution.sample() in not supported

File: sympy/stats/crv.py (hunk 3 above)
Comment (Smit-create) [reply sympy/stats/crv.py:157]: This will require `sympy` to be imported. So, a dictionary could work faster

Comment (czgdp1807) [reply sympy/stats/crv.py:157]: Can you shift `_get_sample_class` dict under this class. It seems faster than `if-else` ladder.

Comment (Smit-create) [reply sympy/stats/crv.py:157]: Then I need to shift the base class below three library classes as the dictionary will produce `NotDefinedError`, so dictionary should be after defining all classes as present

**Comment by Smit-create** [comment]
@czgdp1807  Does this require any further changes?

**Comment by czgdp1807** [comment]
LGTM. Any thoughts @Upabjojr @jmig5776 ?

**Comment by czgdp1807** [comment]
Will be merged tomorrow. Please express your opinions.

**Comment by Upabjojr** [comment]
Just as a minor thought... Maybe separating the code into files by the external library used... But that's just a thought, no need to do it.

**Comment by Upabjojr** [comment]
I don't see the point in removing the class inheritance and implementing it manually with a dictionary. Too much added complexity for no clear gain.

**Comment by jmig5776** [comment]
@czgdp1807 Looks good to me.

File: sympy/stats/crv.py [hunk 6]
```diff
@@ -301,7 +288,7 @@ def __new__(cls, *args):
     def check(*args):
         pass
 
-    def sample(self, size=1, library='scipy'):
+    def sample(self, size=(1,), library='scipy'):
```
[the remark this answers is not in the document]
Comment (Smit-create) [reply sympy/stats/crv.py:282]: So, the default should be `size=()`?

[the remark this answers is not in the document]
Comment (Upabjojr) [reply sympy/stats/crv.py:282]: NumPy (and also SymPy now) considers scalars to be of `size=()`, while `size=(1,)` is an array of one element.

The question is: should it return a scalar or a one-element array?

[the remark this answers is not in the document]
Comment (Smit-create) [reply sympy/stats/crv.py:282]: As the sampling of sympy is based on NumPy/SciPy, I agree that we should follow their convention too. I will update the PR accordingily.

**Comment by Smit-create** [comment]
@Upabjojr  Is this good to go now?

**Comment by czgdp1807** [comment]
Can we merge this by tomorrow?

File: sympy/stats/crv.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Upabjojr) [reply sympy/stats/crv.py]: Are all cases tested in the tests? We should be able to support both `size=()` and `size=(1,)` (returning a scalar and a 1 element array, respectively).

By the way, we should also check if `size=(1, 1)` works, returning a 1-element 2-dimensional array.

[the remark this answers is not in the document]
Comment (Upabjojr) [reply sympy/stats/crv.py]: (there is an `Array` class in `sympy.tensor.array`)

**Comment by Upabjojr** [comment]
> Can we merge this by tomorrow?

If you want, yes.

File: sympy/stats/crv.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Smit-create) [reply sympy/stats/crv.py]: Yes all of them are covered. For, `size=()`, `size=3`, `size=(2,2)`

**Comment by czgdp1807** [comment]
I don't know why but a lot of lines are uncovered, https://codecov.io/gh/sympy/sympy/pull/19342/diff
@Smit-create Can you please check?

**Comment by Smit-create** [comment]
Hi, I have covered all the lines and for all the RVs in the map. The reason here can be because of dependencies on external libraries, the tests might be skipped, so they are not reaching to their respective files. I observed the same with the CRV pr too earlier.

**Comment by Smit-create** [comment]
Locally running for `sympy.stats.crv`,
Screenshot from 2020-05-27 11-09-59

**Comment by czgdp1807** [comment]
Will be merged tomorrow.

**Comment by czgdp1807** [comment]
Let's start discussion of simulating stochastic process, probably using external libraries.

**Pull request merged**